subprojects/file-watching/src/main/java/org/gradle/internal/watch/registry/impl/HierarchicalFileWatcherUpdater.java

```diff
@@ -109,15 +152,48 @@ private void updateWatchedDirectories(Set<Path> newWatchRoots) {
                 .map(Path::toFile)
                 .collect(Collectors.toList())
             );
-            watchedRoots.removeAll(watchRootsToRemove);
+            watchedHierarchies.removeAll(watchRootsToRemove);
         }
         if (!newWatchRoots.isEmpty()) {
             watcher.startWatching(newWatchRoots.stream()
                 .map(Path::toFile)
                 .collect(Collectors.toList())
             );
-            watchedRoots.addAll(newWatchRoots);
+            watchedHierarchies.addAll(newWatchRoots);
+        }
+        LOGGER.info("Watching {} directory hierarchies to track changes", watchedHierarchies.size());
+    }
+
+    private static class RootProjectDirectory {
+        private final Path path;
+        private final String prefix;
+
+        public RootProjectDirectory(Path path) {
+            this.path = path;
+            this.prefix = path.toString() + File.separator;
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) {
+                return true;
+            }
+            if (o == null || getClass() != o.getClass()) {
+                return false;
+            }
+
+            RootProjectDirectory that = (RootProjectDirectory) o;
+
+            return prefix.equals(that.prefix);
+        }
+
+        @Override
+        public int hashCode() {
+            return prefix.hashCode();
+        }
+
+        public boolean contains(String shouldWatchDirectoryPathString) {
+            return shouldWatchDirectoryPathString.startsWith(prefix);
```

Comment:
When calling this method, we are calling `Path.toString()` in the only place it's called from. Do you think this is still faster than calling `Path.startsWith()`? An advantage of using the latter would be that we could potentially replace `RootProjectDirectory` with a simple `Path` object.